src/client/components/Navigation/LanguageSettings.js

```diff
@@ -0,0 +1,177 @@
+import { saveSettings, setLocale } from '../../settings/settingsActions';
+import { getIsAuthenticated, getLocale } from '../../reducers';
+import PopoverMenu, { PopoverMenuItem } from '../PopoverMenu/PopoverMenu';
+import Loading from '../Icon/Loading';
+import './LanguageSettings.less';
+
+@connect(
+  state => ({
+    authenticated: getIsAuthenticated(state),
+    locale: getLocale(state),
+  }),
+  {
+    saveSettings,
+    setLocale,
+  },
+)
+class LanguageSettings extends React.Component {
+  static propTypes = {
+    authenticated: PropTypes.bool.isRequired,
+    saveSettings: PropTypes.func.isRequired,
+    setLocale: PropTypes.func.isRequired,
+    locale: PropTypes.string,
+  };
+
+  static defaultProps = {
+    locale: 'en-US',
+  };
+
+  static getValidLocale = locale => {
+    const availableLocales = _.keys(availableLocalesToReactIntl);
+    let validLocale = 'en-US';
+
+    for (let i = 0; i < availableLocales.length; i += 1) {
+      const currentLocale = availableLocales[i];
+      const currentLocaleShortValue = availableLocalesToReactIntl[currentLocale];
+
+      if (locale === currentLocale) {
+        validLocale = currentLocale;
+        break;
+      } else if (locale === currentLocaleShortValue) {
+        validLocale = currentLocale;
+        break;
+      }
+    }
+    return validLocale;
+  };
+
+  constructor(props) {
+    super(props);
+
+    let localeStorageLanguage;
+    let browserLanguage;
+
+    if (localStorage) localeStorageLanguage = localStorage.language;
+    if (navigator) {
+      browserLanguage = LanguageSettings.getValidLocale(
+        _.get(navigator.languages, 0, navigator.language),
+      );
+    }
+
+    const localLanguage = localeStorageLanguage || browserLanguage;
+    const selectedLanguage = props.authenticated ? props.locale : localLanguage;
+
+    this.state = {
+      languageSettingsVisible: false,
+      selectedLanguage,
+      loading: false,
+      selectedLoadingLanguage: '',
+    };
+
+    this.handleLanguageSettingsVisibleChange = this.handleLanguageSettingsVisibleChange.bind(this);
+    this.handleLanguageSettingsSelect = this.handleLanguageSettingsSelect.bind(this);
+  }
+
+  componentDidMount() {
+    this.props.setLocale(this.state.selectedLanguage);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const diffLocale = this.props.locale !== nextProps.locale;
+    if (diffLocale) {
+      this.setState({ selectedLanguage: LanguageSettings.getValidLocale(nextProps.locale) });
+    }
+  }
+
+  setLoadingLanguage(loading, selectedLoadingLanguage) {
+    this.setState({ loading, selectedLoadingLanguage });
+  }
+
+  handleLanguageSettingsVisibleChange(visible) {
+    this.setState({ languageSettingsVisible: visible });
+  }
+
+  handleLanguageSettingsSelect(selectedLanguage) {
+    const { authenticated } = this.props;
+    const { loading } = this.state;
+
+    if (loading) return;
+
+    this.setLoadingLanguage(true, selectedLanguage);
```

Comment:
If use is not authenticated it will have no effect (if user is not authenticated it will set loading to false right away). Might make sense to move it inside `authenticated` branch.